### PR TITLE
Prototype a stylelint shim

### PR DESCRIPTION
**Pull request by jesstelford**
**tl;dr**: I wrote a shim which should simplify future work to improve the coverage tool. It'll help with code maintenance _today_, with the goal of making our coverage more accurate and the `stylelint-polaris` tool more useful _in the future_. [See the migration template for a minimal usage example](https://github.com/Shopify/polaris/blob/prototype-stylelint-shim/polaris-migrator/templates/sass-migration/%7B%7BkebabCase%20migrationName%7D%7D/%7B%7BkebabCase%20migrationName%7D%7D.ts.hbs).

---

In conversation with @aaronccasanova yesterday, he pointed out that except for some extreme edge cases, all our migrations follow an identical pattern of:

1) Check if a given CSS line is something the migration cares about ('spacing' doesn't care about `display:` declarations, for example)
2) Identify the "failure"
3) Try to fix it
4) Note when we can't automatically fix it.

In seeing that pattern, he's been able to simplify logic and converge on a nice code structure as [exemplified by the ](https://github.com/Shopify/polaris/blob/main/polaris-migrator/src/migrations/replace-spacing-lengths/replace-spacing-lengths.ts)`replace-spacing-lengths`[ migration](https://github.com/Shopify/polaris/blob/main/polaris-migrator/src/migrations/replace-spacing-lengths/replace-spacing-lengths.ts).

I began shoring up the `motion` migration to follow this structure and started to see that @aaronccasanova's structure could be extended further to remove duplication and make every migration even more consistent + easy to maintain.

At the same time, I've been thinking on how our coverage tooling (powered by `stylelint`) and migration tooling (powered by `postcss`) relate, and specifically how they're starting to diverge, and wondering what we can do about it. I discovered that `stylelint` supports running "migrations" via its built-in `--fix` option. I immediately thought of a future where we used `stylelint` to execute our `postcss` migrations.

It turns out `stylelint` is actually powered by `postcss`! And the the way Stylelint rules function is:

1) Check if a given CSS line is something the lint rule cares about ('spacing' doesn't care about `display:` declarations, for example)
2) Identify any linting "errors" or "warnings"
3) Optionally; Try to fix it when `--fix` is passed

Which is almost identical to @aaronccasanova's observation and subsequent code structure 🤯 

So, much like Alice, I tumbled down the Rabbit Hole :rabbit-hole: and realised that with just a little shim function, we could write our `postcss` migrations _today_ such that they're compatible with being run within `stylelint` in the future, while ensuring the migrations are consistent and easy to maintain ✅

I'd love to hear thoughts! Is this useful? Is it confusing? Am I aiming in the right direction?

### CODE REVIEW

File: polaris-migrator/src/migrations/replace-spacing-lengths/tests/replace-spacing-lengths.output.scss [hunk 1]
```diff
@@ -23,55 +25,65 @@
...
+  // error: Non-tokenizable value '10px'
+  // error: Non-tokenizable value '10rem'
   padding: 10px 10rem;
   // polaris-migrator: Unable to migrate the following expression. Please upgrade manually.
+  // error: Non-tokenizable value '10rem'
   // padding: var(--p-space-4) 10rem;
   padding: 16px 10rem;
   // polaris-migrator: Unable to migrate the following expression. Please upgrade manually.
+  // error: Non-tokenizable value '10px'
   // padding: 10px var(--p-space-4);
```
Comment (jesstelford) [inline polaris-migrator/src/migrations/replace-spacing-lengths/tests/replace-spacing-lengths.output.scss:57]: Still outputs the partial fix where possible, but enhances that with the specific reason it couldn't be fully fixed.

Huge DX win!

File: polaris-migrator/src/migrations/replace-spacing-lengths/tests/replace-spacing-lengths.output.scss [hunk 2]
```diff
@@ -23,55 +25,65 @@
   padding: var(--p-space-4, 16px);
   // Comment
   // polaris-migrator: Unable to migrate the following expression. Please upgrade manually.
-  // padding: 10px;
+  // error: Non-tokenizable value '10px'
```
Comment (jesstelford) [inline polaris-migrator/src/migrations/replace-spacing-lengths/tests/replace-spacing-lengths.output.scss:28]: These error messages will be the exact same as exposed by `stylelint` once we migrate there. 🎉

File: polaris-migrator/src/migrations/replace-spacing-lengths/replace-spacing-lengths.ts [hunk 3]
```diff
@@ -26,87 +24,81 @@ export default function replaceSpacingLengths(
...
-      if (targets.some(({replaced}) => !replaced || hasNumericOperator)) {
-        decl.before(
-          createInlineComment(POLARIS_MIGRATOR_COMMENT, {prose: true}),
-        );
-        decl.before(
-          createInlineComment(`${decl.prop}: ${parsedValue.toString()};`),
-        );
-      } else {
-        decl.value = parsedValue.toString();
-      }
```
Comment (jesstelford) [inline polaris-migrator/src/migrations/replace-spacing-lengths/replace-spacing-lengths.ts:65]: Begone repeated boilerplate code! 🪄

**Comment by BPScott** [comment]
I think my question is: In what cases would we want migration checks and fixes to exist in perpetuity?

Stylelint rules hang around forever - they spot mistakes that could be made and fix them. However the modifications in migrations tend to be one and done. You apply a migration like "replace usage of `spacing()` with tokens" and then you delete the `spacing()` function from your codebase as it is no longer used. At that point if you ever try to use the spacing() function it is an error because that function no longer exists. `Web` already enables the [`scss/no-global-function-names`](https://github.com/stylelint-scss/stylelint-scss/blob/master/src/rules/no-global-function-names/README.md) lint rule that warns against doing `spacing()` (which would compile but silently put `spacing()` in your compiled scss), and because the function no longer exists `legacy-polaris-v8.spacing()` would result in a compile time error. What extra value does a lint rule that says "you can't use spacing()" provide that is not already covered by that status quo?


That said, if the plan is "We stop using codemod for migrations, instead these fixes are exposed as a stylelint rules that contains autofixes where appropriate" then that sounds legit.

**Comment by jesstelford** [comment]
@BPScott Yes, this is a fantastic question!

There's really two distinct times a developer might want their code automatically fixed for them:

1. When working in their IDE, and an auto-fix suggestion says _"Hey, that could be a Polaris token. Click to fix."_
2. When doing Polaris major version migrations.

For #1, we have `stylelint-polaris`, which implements some basic warnings, but cannot fix them.

For #2, we have `polaris-migrator`, which can fix issues based on more complex logic than the `stylelint-polaris`, but cannot be used in an IDE.

These two use-cases are distinct, but it turns out they can use the exact same tech stack: `stylelint` rules. This allows detection _and_ fixing whether it's from an IDE or from a `migrator` cli.

The "one and done" migrations would be separate rules (bundled into `8-to-9` style presets) compared to the "fix this in my IDE" rules. Stylelint also allows executing rules from within other rules, so a migration may leverage the same rules as an IDE one, or vice versa.

All of that is future-looking. This PR is mearly the first step in that direction (one I hope I'm aligned with the rest of the team on 😅)

File: polaris-migrator/src/utilities/sass.ts [hunk 4]
```diff
@@ -251,3 +263,398 @@ export function createInlineComment(text: string, options?: {prose?: boolean}) {
...
+        ruleWalker: (decl: PostCSSRule) => false | void,
+      ) => void;
+    };
+  },
+  context: PluginContext,
+) => (root: Root) => void;
+
+// Expose a stylelint-like API for creating sass migrators so we can easily
+// migrate to that tool in the future.
+function convertStylelintRuleToPostcssProcessor(ruleFn: StylelintRule) {
```
Comment (jesstelford) [inline polaris-migrator/src/utilities/sass.ts:344]: This is the shim function.

Approx 90% of this function would be deleted by migrating to `stylelint`, meaning this function is just a shim to get `postcss`'s API to look like `stylelint`'s (for now).

File: polaris-migrator/src/utilities/sass.ts [hunk 5]
```diff
@@ -251,3 +263,398 @@ export function createInlineComment(text: string, options?: {prose?: boolean}) {
...
+          // Normally, this comes from stylelint config, but for this shim we
+          // just hard-code it, and instead rely on the "seconary" options
+          // object for passing through the jscodeshift options.
+          true,
+          {
+            ...options,
+            partialFixStrategy: 'report',
+            __injectReportsAsComments: true,
+          },
+          {fix: true},
```
Comment (jesstelford) [inline polaris-migrator/src/utilities/sass.ts:376]: The `fix` option is passed by `stylelint` when the cli switch `--fix` is passed. Since this is within our migrator cli, we hard-code it to `true` for this shim.

File: polaris-migrator/src/utilities/sass.ts [hunk 6]
```diff
@@ -251,3 +263,398 @@ export function createInlineComment(text: string, options?: {prose?: boolean}) {
...
+        // rules accidentally retaining scoped variables, etc.
+        const ruleProcessor = ruleFn(
+          // Normally, this comes from stylelint config, but for this shim we
+          // just hard-code it, and instead rely on the "seconary" options
+          // object for passing through the jscodeshift options.
+          true,
+          {
+            ...options,
+            partialFixStrategy: 'report',
+            __injectReportsAsComments: true,
```
Comment (jesstelford) [inline polaris-migrator/src/utilities/sass.ts:374]: This is an example of receiving options via `stylelint` rule config. In this case, we're instructing the migrations to insert comments (because that's the only way to get output from `postcss`), but it could also be set to emit `stylelint` "reports" which are then consumable by other tools such as LSPs, VSCode, etc.

File: polaris-migrator/src/utilities/sass.ts [hunk 7]
```diff
@@ -251,3 +263,398 @@ export function createInlineComment(text: string, options?: {prose?: boolean}) {
...
+      }
+
+      const reportsForNode = reports.get(newReport.node)!;
+
+      if (
+        reportsForNode.findIndex(
+          (existingReport) =>
+            existingReport.severity === newReport.severity &&
+            existingReport.message === newReport.message,
+        ) === -1
```
Comment (jesstelford) [inline polaris-migrator/src/utilities/sass.ts:414]: Deduping reports to avoid noise. Particularly useful for example `padding: 10px 10px 10px 10px` - we only want 1 warning, not 4 identical ones.

File: polaris-migrator/src/utilities/sass.ts [hunk 8]
```diff
@@ -251,3 +263,398 @@ export function createInlineComment(text: string, options?: {prose?: boolean}) {
...
+      },
+    };
+
+    return postcss(plugin).process(fileInfo.source, {
+      syntax: require('postcss-scss'),
+    }).css;
+  };
+}
+
+export function createStylelintRule(ruleName: string, ruleFn: PolarisMigrator) {
```
Comment (jesstelford) [inline polaris-migrator/src/utilities/sass.ts:389]: This factory function matches the `stylelint` API as closely as possible. There are only a couple of lines in this function which would be removed once we move to `stylelint`.

File: polaris-migrator/src/utilities/sass.ts [hunk 9]
```diff
@@ -251,3 +263,387 @@ export function createInlineComment(text: string, options?: {prose?: boolean}) {
...
+      serialiseSuggestion: (node: T) => string;
+    }): (node: T) => false | void;
+    function createWalker<T extends PostCSSNode>(args: {
+      valueParserKey?: keyof T;
+      walker:
+        | ((node: T) => false | void)
+        | ((node: T, parsedValue: ParsedValue) => false | void);
+      mutableKeys?: (keyof T)[];
+      serialiseSuggestion: (node: T) => string;
+    }): (node: T) => false | void {
```
Comment (jesstelford) [inline polaris-migrator/src/utilities/sass.ts:473]: Typescript 😭

File: polaris-migrator/src/utilities/sass.ts [hunk 10]
```diff
@@ -251,3 +263,387 @@ export function createInlineComment(text: string, options?: {prose?: boolean}) {
...
+      mutableKeys?: (keyof T)[];
+      serialiseSuggestion: (node: T) => string;
+    }): (node: T) => false | void;
+    function createWalker<T extends PostCSSNode>(args: {
+      valueParserKey: keyof T;
+      walker: (node: T, parsedValue: ParsedValue) => false | void;
+      mutableKeys?: (keyof T)[];
+      serialiseSuggestion: (node: T) => string;
+    }): (node: T) => false | void;
+    function createWalker<T extends PostCSSNode>(args: {
```
Comment (jesstelford) [inline polaris-migrator/src/utilities/sass.ts:466]: This factory function wraps up common logic across all of our migrators:

- Tracking if changes have occurred to a postcss Node (previously done manually in each migration via a `targets` array)
- Adding comments if there's a partial migration
- Add comments when there's an error/warning detected, but it's unmigratable

File: polaris-migrator/src/utilities/sass.ts [hunk 11]
```diff
@@ -251,3 +263,387 @@ export function createInlineComment(text: string, options?: {prose?: boolean}) {
...
+      {
+        options: secondaryOptions,
+        methods: {
+          report: addDedupedReport,
+          each,
+          walk,
+          walkAtRules,
+          walkComments,
+          walkDecls,
+          walkRules,
```
Comment (jesstelford) [inline polaris-migrator/src/utilities/sass.ts:641]: These util methods cover most of the [postcss `Root` API](https://postcss.org/api/#root) that our migrations will/might use.

**Comment by sam-b-rose** [comment]
This is an excellent exploration aligning our migration structure with how Stylelint structures rules. This iteration does bring the abstraction closer to Stylelint's API, but it also introduces a lot of complexity and changes to our current migrations–even if only temporary.

Until we are ready and sure we'd like to migrate the migrations to our Stylelint plugin, I'd like to let this work simmer if that is alright 🫕. This will certainly be valuable in helping form our strategy for integrating `stylelint-polaris` with Shopify/web :+1:

**Comment by jesstelford** [comment]
Fair call: This PR has a lot going on. I'll split out a couple things I think will be useful for us to have in separate PRs 👍
